File: golang/widget-store/handlers.go
package main

import (
	"context"
	"log/slog"
	"net/http"
	"os"
	"strconv"
	"time"

	"github.com/dbos-inc/dbos-transact-golang/dbos"
	"github.com/gin-gonic/gin"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

const PAYMENT_STATUS = "payment_status"
const PAYMENT_ID = "payment_id"
const ORDER_ID = "order_id"

// HTTP handlers
func getProduct(c *gin.Context, db *pgxpool.Pool, logger *slog.Logger) {
	var product Product
	err := db.QueryRow(context.Background(),
		"SELECT product_id, product, description, inventory, price FROM products LIMIT 1").
		Scan(&product.ProductID, &product.Product, &product.Description, &product.Inventory, &product.Price)

	if err != nil {
		logger.Error("product query failed", "error", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch product"})
		return
	}
	c.JSON(http.StatusOK, product)
}

func getOrders(c *gin.Context, db *pgxpool.Pool, logger *slog.Logger) {
	rows, err := db.Query(context.Background(),
		"SELECT order_id, order_status, last_update_time, progress_remaining FROM orders")
	if err != nil {
		logger.Error("orders database query failed", "error", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch orders"})
		return
	}
	defer rows.Close()

	orders := []Order{}
	for rows.Next() {
		var order Order
		err := rows.Scan(&order.OrderID, &order.OrderStatus, &order.LastUpdateTime, &order.ProgressRemaining)
		if err != nil {
			logger.Error("order data parsing failed", "error", err)
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to process orders"})
			return
		}
		orders = append(orders, order)
	}

	c.JSON(http.StatusOK, orders)
}

func getOrder(c *gin.Context, db *pgxpool.Pool, logger *slog.Logger) {
	idStr := c.Param("id")
	id, err := strconv.Atoi(idStr)
	if err != nil {
		logger.Warn("invalid order ID", "error", err, "id", idStr)
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid order ID"})
		return
	}

	var order Order
	err = db.QueryRow(context.Background(),
		"SELECT order_id, order_status, last_update_time, progress_remaining FROM orders WHERE order_id = $1", id).
		Scan(&order.OrderID, &order.OrderStatus, &order.LastUpdateTime, &order.ProgressRemaining)

	if err != nil {
		if err == pgx.ErrNoRows {
			logger.Warn("order not found", "order", id)
			c.JSON(http.StatusNotFound, gin.H{"error": "Order not found"})
		} else {
			logger.Error("order database query failed", "error", err, "order", id)
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch order"})
		}
		return
	}
	c.JSON(http.StatusOK, order)
}

func restock(c *gin.Context, db *pgxpool.Pool, logger *slog.Logger) {
	_, err := db.Exec(context.Background(), "UPDATE products SET inventory = 100")
	if err != nil {
		logger.Error("inventory update failed", "error", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to restock inventory"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Restocked successfully"})
}

// XXX can we fold our context inside the gin context?
// and more generally, how do funky contexts play together
func checkoutEndpoint(c *gin.Context, dbosCtx dbos.DBOSContext, logger *slog.Logger) {
	idempotencyKey := c.Param("idempotency_key")

	// Start the checkout workflow with the idempotency key
	_, err := dbos.RunWorkflow(dbosCtx, checkoutWorkflow, "", dbos.WithWorkflowID(idempotencyKey))
	if err != nil {
		logger.Error("checkout workflow start failed", "error", err, "key", idempotencyKey)
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Checkout failed to start"})
		return
	}

	payment_id, err := dbos.GetEvent[string](dbosCtx, idempotencyKey, PAYMENT_ID, 60*time.Second)
	if err != nil || payment_id == "" {
		logger.Error("payment ID retrieval failed", "key", idempotencyKey)
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Checkout failed"})
		return
	}

	c.String(http.StatusOK, payment_id)
}

func paymentEndpoint(c *gin.Context, dbosCtx dbos.DBOSContext, logger *slog.Logger) {
	paymentID := c.Param("payment_id")
	paymentStatus := c.Param("payment_status")

	err := dbos.Send(dbosCtx, paymentID, paymentStatus, PAYMENT_STATUS)
	if err != nil {
		logger.Error("payment notification failed", "error", err, "payment", paymentID, "status", paymentStatus)
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to process payment"})
		return
	}

	orderID, err := dbos.GetEvent[string](dbosCtx, paymentID, ORDER_ID, 60*time.Second)
	if err != nil || orderID == "" {
		logger.Error("order ID retrieval failed", "payment", paymentID)
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Payment failed to process"})
		return
	}

	c.String(http.StatusOK, orderID)
}

func crashApplication(c *gin.Context, logger *slog.Logger) {
	logger.Warn("application crash requested")
	c.JSON(http.StatusOK, gin.H{"message": "Crashing application..."})
	// Give time for response to be sent
	go func() {
		time.Sleep(100 * time.Millisecond)
		logger.Error("intentional crash for demo")
		os.Exit(1)
	}()
}

File: rust/widget-store/src/handlers.rs
//! HTTP handlers mirroring the Go widget-store's endpoints.

use crate::store;
use crate::workflows::{ORDER_ID, PAYMENT_ID, PAYMENT_STATUS};
use axum::{
    extract::{Path, State},
    http::StatusCode,
    response::{Html, IntoResponse, Response},
    Json,
};
use durare::{DurableEngine, WorkflowOptions};
use serde_json::json;
use std::sync::Arc;
use std::time::Duration;

const EVENT_TIMEOUT: Duration = Duration::from_secs(60);

fn error_json(status: StatusCode, msg: &str) -> Response {
    (status, Json(json!({ "error": msg }))).into_response()
}

pub async fn index() -> Html<&'static str> {
    Html(include_str!("../html/app.html"))
}

pub async fn get_product() -> Response {
    match store::get_product().await {
        Ok(p) => Json(p).into_response(),
        Err(e) => {
            eprintln!("product query failed: {e}");
            error_json(StatusCode::INTERNAL_SERVER_ERROR, "Failed to fetch product")
        }
    }
}

pub async fn get_orders() -> Response {
    match store::get_orders().await {
        Ok(orders) => Json(orders).into_response(),
        Err(e) => {
            eprintln!("orders query failed: {e}");
            error_json(StatusCode::INTERNAL_SERVER_ERROR, "Failed to fetch orders")
        }
    }
}

pub async fn get_order(Path(id): Path<String>) -> Response {
    let Ok(id) = id.parse::<i32>() else {
        return error_json(StatusCode::BAD_REQUEST, "Invalid order ID");
    };
    match store::get_order(id).await {
        Ok(Some(order)) => Json(order).into_response(),
        Ok(None) => error_json(StatusCode::NOT_FOUND, "Order not found"),
        Err(e) => {
            eprintln!("order query failed: {e}");
            error_json(StatusCode::INTERNAL_SERVER_ERROR, "Failed to fetch order")
        }
    }
}

pub async fn restock() -> Response {
    match store::restock().await {
        Ok(()) => Json(json!({ "message": "Restocked successfully" })).into_response(),
        Err(e) => {
            eprintln!("restock failed: {e}");
            error_json(
                StatusCode::INTERNAL_SERVER_ERROR,
                "Failed to restock inventory",
            )
        }
    }
}

/// Start (or, on retry, attach to) the checkout workflow under the caller's
/// idempotency key, then wait for it to publish the payment id.
pub async fn checkout(
    State(engine): State<Arc<DurableEngine>>,
    Path(idempotency_key): Path<String>,
) -> Response {
    let started = engine
        .start::<String, String>(
            "checkoutWorkflow",
            String::new(),
            WorkflowOptions::with_id(&idempotency_key),
        )
        .await;
    if let Err(e) = started {
        eprintln!("checkout start failed for {idempotency_key}: {e}");
        return error_json(
            StatusCode::INTERNAL_SERVER_ERROR,
            "Checkout failed to start",
        );
    }

    match engine
        .get_event::<String>(&idempotency_key, PAYMENT_ID, EVENT_TIMEOUT)
        .await
    {
        Ok(Some(payment_id)) if !payment_id.is_empty() => payment_id.into_response(),
        _ => {
            eprintln!("payment id retrieval failed for {idempotency_key}");
            error_json(StatusCode::INTERNAL_SERVER_ERROR, "Checkout failed")
        }
    }
}

/// The payment processor's webhook: deliver the status to the waiting checkout
/// workflow, then wait for it to publish the order id.
pub async fn payment_webhook(
    State(engine): State<Arc<DurableEngine>>,
    Path((payment_id, payment_status)): Path<(String, String)>,
) -> Response {
    if let Err(e) = engine
        .send(&payment_id, payment_status, PAYMENT_STATUS)
        .await
    {
        eprintln!("payment notification failed for {payment_id}: {e}");
        return error_json(
            StatusCode::INTERNAL_SERVER_ERROR,
            "Failed to process payment",
        );
    }

    match engine
        .get_event::<String>(&payment_id, ORDER_ID, EVENT_TIMEOUT)
        .await
    {
        Ok(Some(order_id)) if !order_id.is_empty() => order_id.into_response(),
        _ => {
            eprintln!("order id retrieval failed for {payment_id}");
            error_json(
                StatusCode::INTERNAL_SERVER_ERROR,
                "Payment failed to process",
            )
        }
    }
}

/// Crash the process — the point of the demo: watch the checkout and dispatch
/// workflows resume exactly where they stopped.
pub async fn crash_application() -> Response {
    tokio::spawn(async {
        // Give the response time to flush.
        tokio::time::sleep(Duration::from_millis(100)).await;
        std::process::exit(1);
    });
    Json(json!({ "message": "Crashing application..." })).into_response()
}
